src/ws.nim

```diff
@@ -205,226 +280,361 @@ proc encodeFrame(f: Frame): seq[byte] =
+  if ws.masked:
+    maskKey = genMaskKey(ws.rng)
+
+  if opcode notin {Opcode.Text, Opcode.Cont, Opcode.Binary}:
+    discard await ws.tcpSocket.write(encodeFrame(Frame(
+        fin: true,
+        rsv1: false,
+        rsv2: false,
+        rsv3: false,
+        opcode: opcode,
+        mask: ws.masked,
+        data: data, # allow sending data with close messages
+        maskKey: maskKey)))
 
-  var frame: Frame
-  # Read the flags and fin from the header.
+    return
 
-  var hf = cast[HeaderFlags](b0 shr 4)
-  frame.fin = fin in hf
-  frame.rsv1 = rsv1 in hf
-  frame.rsv2 = rsv2 in hf
-  frame.rsv3 = rsv3 in hf
+  let maxSize = ws.frameSize
+  # Send stuff in 1 megabyte chunks to prevent IOErrors.
+  # This really large packets.
+  var i = 0
+  while i < data.len:
+    let len = min(data.len, (maxSize + i))
+    let inFrame = Frame(
+        fin: if (i + len >= data.len): true else: false,
+        rsv1: false,
+        rsv2: false,
+        rsv3: false,
+        opcode: if i > 0: Opcode.Cont else: opcode, # fragments have to be `Continuation` frames
+        mask: ws.masked,
+        data: data[i ..< len],
+        maskKey: maskKey)
+
+    discard await ws.tcpSocket.write(encodeFrame(inFrame))
+    i += len
+
+proc send*(ws: WebSocket, data: string): Future[void] =
+  send(ws, toBytes(data), Opcode.Text)
+
+proc handleClose*(ws: WebSocket, frame: Frame) {.async.} =
+  logScope:
+    fin = frame.fin
+    masked = frame.mask
+    opcode = frame.opcode
+    serverState = ws.readyState
+
+  debug "Handling close sequence"
+  if ws.readyState == ReadyState.Open or ws.readyState == ReadyState.Closing:
+    # Read control frame payload.
+    var data = newSeq[byte](frame.length)
+    if frame.length > 0:
+      # Read the data.
+      await ws.tcpSocket.readExactly(addr data[0], int frame.length)
+      unmask(data.toOpenArray(0, data.high), frame.maskKey)
+
+    var code: Status
+    if data.len > 0:
+      let ccode = uint16.fromBytesBE(data[0..<2]) # first two bytes are the status
+      doAssert(ccode > 999, "No valid code in close message!")
+      code = Status(ccode)
+      data = data[2..data.high]
+
+    var rcode = Status.Fulfilled
+    var reason = ""
+    if not isNil(ws.onClose):
+      try:
+        (rcode, reason) = ws.onClose(code, string.fromBytes(data))
+      except CatchableError as exc:
+        debug "Exception in Close callback, this is most likelly a bug", exc = exc.msg
+
+    # don't respong to a terminated connection
+    if ws.readyState != ReadyState.Closing:
+      await ws.send(prepareCloseBody(rcode, reason), Opcode.Close)
+
+    await ws.tcpSocket.closeWait()
+    ws.readyState = ReadyState.Closed
+  else:
+    raiseAssert("Invalid state during close!")
 
-  var opcode = b0 and 0x0f
-  if opcode notin WSOpCode:
-    raise newException(WebSocketError, "Unexpected  websocket opcode")
-  frame.opcode = (opcode).Opcode
+proc handleControl*(ws: WebSocket, frame: Frame) {.async.} =
+  ## handle control frames
```

Comment:
We can add a check on the size of the control frame data. According to the [spec](https://tools.ietf.org/html/rfc6455#section-5.5.1):

> All control frames MUST have a payload length of 125 bytes or less

Reply:
Ah, nice catch.